spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/config/s3/S3ConfigDataLocationResolver.java

```diff
@@ -0,0 +1,131 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.awspring.cloud.autoconfigure.config.s3;
+
+import io.awspring.cloud.autoconfigure.config.AbstractAwsConfigDataLocationResolver;
+import io.awspring.cloud.autoconfigure.core.*;
+import io.awspring.cloud.autoconfigure.s3.AwsS3ClientCustomizer;
+import io.awspring.cloud.autoconfigure.s3.S3KeysMissingException;
+import io.awspring.cloud.autoconfigure.s3.properties.S3Properties;
+import io.awspring.cloud.s3.crossregion.CrossRegionS3Client;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.apache.commons.logging.Log;
+import org.springframework.boot.BootstrapContext;
+import org.springframework.boot.BootstrapRegistry;
+import org.springframework.boot.context.config.ConfigDataLocation;
+import org.springframework.boot.context.config.ConfigDataLocationNotFoundException;
+import org.springframework.boot.context.config.ConfigDataLocationResolverContext;
+import org.springframework.boot.context.config.Profiles;
+import org.springframework.boot.context.properties.bind.Bindable;
+import org.springframework.boot.context.properties.bind.Binder;
+import org.springframework.boot.logging.DeferredLogFactory;
+import org.springframework.util.ClassUtils;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.S3ClientBuilder;
+
+/**
+ * Resolves config data locations in AWS S3.
+ *
+ * @author Kunal Varpe
+ * @author Matej Nedic
+ * @since 3.1.0
+ */
+public class S3ConfigDataLocationResolver extends AbstractAwsConfigDataLocationResolver<S3ConfigDataResource> {
+
+	/**
+	 * AWS S3 Config Data prefix.
+	 */
+	public static final String PREFIX = "aws-s3:";
+
+	private final Log log;
+
+	public S3ConfigDataLocationResolver(DeferredLogFactory deferredLogFactory) {
+		this.log = deferredLogFactory.getLog(S3ConfigDataLocationResolver.class);
+	}
+
+	@Override
+	protected String getPrefix() {
+		return PREFIX;
+	}
+
+	@Override
+	public List<S3ConfigDataResource> resolveProfileSpecific(ConfigDataLocationResolverContext resolverContext,
+			ConfigDataLocation location, Profiles profiles) throws ConfigDataLocationNotFoundException {
+
+		S3Properties s3Properties = loadProperties(resolverContext.getBinder());
+
+		registerBean(resolverContext, AwsProperties.class, loadAwsProperties(resolverContext.getBinder()));
+		registerBean(resolverContext, S3Properties.class, s3Properties);
+		registerBean(resolverContext, CredentialsProperties.class,
+				loadCredentialsProperties(resolverContext.getBinder()));
+		registerBean(resolverContext, RegionProperties.class, loadRegionProperties(resolverContext.getBinder()));
+
+		registerAndPromoteBean(resolverContext, S3Client.class, BootstrapRegistry.InstanceSupplier.of(
```

Comment:
If it is a supplier, perhaps it is better to pass a lambda there? 

```
BootstrapRegistry.InstanceSupplier.of(() -> ...);
```

Reply:
This has been addressed